hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/hudi/IncrementalRelation.scala

```diff
@@ -157,27 +161,40 @@ class IncrementalRelation(val sqlContext: SQLContext,
     } else {
       log.info("Additional Filters to be applied to incremental source are :" + filters)
 
-      var df: DataFrame = sqlContext.createDataFrame(sqlContext.sparkContext.emptyRDD[Row], usedSchema)
+      var prunedSchema = StructType(Seq())
+      if (!requiredColumns.contains(HoodieRecord.COMMIT_TIME_METADATA_FIELD)) {
+        prunedSchema = prunedSchema.add(usedSchema(HoodieRecord.COMMIT_TIME_METADATA_FIELD))
+      }
+      requiredColumns.foreach(col => {
+        val field = usedSchema.find(_.name == col)
+        if (field.isDefined) {
+          prunedSchema = prunedSchema.add(field.get)
+        }
+      })
+      var df: DataFrame = sqlContext.createDataFrame(sqlContext.sparkContext.emptyRDD[Row], prunedSchema)
 
       if (metaBootstrapFileIdToFullPath.nonEmpty) {
         df = sqlContext.sparkSession.read
                .format("hudi")
-               .schema(usedSchema)
+               .schema(prunedSchema)
                .option(DataSourceReadOptions.READ_PATHS.key, filteredMetaBootstrapFullPaths.mkString(","))
                .load()
       }
 
       if (regularFileIdToFullPath.nonEmpty)
       {
         df = df.union(sqlContext.read.options(sOpts)
-                        .schema(usedSchema)
+                        .schema(prunedSchema)
                         .parquet(filteredRegularFullPaths.toList: _*)
                         .filter(String.format("%s >= '%s'", HoodieRecord.COMMIT_TIME_METADATA_FIELD,
                           commitsToReturn.head.getTimestamp))
                         .filter(String.format("%s <= '%s'", HoodieRecord.COMMIT_TIME_METADATA_FIELD,
                           commitsToReturn.last.getTimestamp)))
       }
 
+      if (!requiredColumns.contains(HoodieRecord.COMMIT_TIME_METADATA_FIELD)) {
```

Comment:
Because IncrementalRelation extends PrunedScan here, the SparkSQL engine passes the required columns to IncrementalRelation and it expects the returned RDD only to have the required columns.
If we don't remove the non-required column here, it will cause schema mismatch errors.